src/main/java/com/reservation/payment/domain/Payment.java

```diff
@@ -0,0 +1,116 @@
+    private PaymentCompletedAt completedAt;
+    private PaymentFailedAt failedAt;
+    private FailureReason failureReason;
+
+    private final List<PaymentEvent> events = new ArrayList<>();
+
+    private Payment(PaymentId paymentId, ReservationId reservationId, Money amount,
+                    PaymentMethod method, PaymentInitiatedAt initiatedAt) {
+        this.paymentId = Objects.requireNonNull(paymentId);
+        this.reservationId = Objects.requireNonNull(reservationId);
+        this.amount = Objects.requireNonNull(amount);
+        this.method = Objects.requireNonNull(method);
+        this.initiatedAt = Objects.requireNonNull(initiatedAt);
+        this.status = PaymentStatus.INITIATED;
+    }
+
+    public static Payment create(PaymentId id, ReservationId reservationId, Money amount,
+                                   PaymentMethod method, PaymentInitiatedAt initiatedAt) {
+        return new Payment(id, reservationId, amount, method, initiatedAt);
+    }
+
+    public PaymentId getPaymentId() {
+        return paymentId;
+    }
+
+    public ReservationId getReservationId() {
+        return reservationId;
+    }
+
+    public boolean hasSameReservation(ReservationId id) {
+        return this.reservationId.equals(id);
+    }
+
+    public void markAsCompleted(PaymentCompletedAt completedAt) {
+        requireNotCompletedOrFailed();
+        this.completedAt = Objects.requireNonNull(completedAt);
+        this.status = PaymentStatus.COMPLETED;
+        registerEvent(new PaymentCompletedEvent(paymentId, reservationId, completedAt));
+    }
+
+    public void markAsFailed(PaymentFailedAt failedAt, FailureReason reason) {
+        requireNotCompletedOrFailed();
+        this.failedAt = Objects.requireNonNull(failedAt);
+        this.failureReason = Objects.requireNonNull(reason);
+        this.status = PaymentStatus.FAILED;
+        registerEvent(new PaymentFailedEvent(paymentId, reservationId, reason, failedAt));
+    }
+
+    private void requireNotCompletedOrFailed() {
+        if (isCompleted()) {
+            throw new PaymentAlreadyCompletedException();
+        }
+        if (isFailed()) {
+            throw new PaymentAlreadyFailedException();
+        }
+    }
+
+    private boolean isCompleted() {
+        return status == PaymentStatus.COMPLETED;
+    }
+
+    private boolean isFailed() {
+        return status == PaymentStatus.FAILED;
+    }
+
+    private boolean isInitiated() {
+        return status == PaymentStatus.INITIATED;
+    }
+
+    public PaymentSnapshot describeSnapshot() {
+        return new PaymentSnapshot(
+                paymentId,
+                reservationId,
+                amount,
+                method,
+                status,
+                initiatedAt,
+                completedAt,
+                failedAt,
+                failureReason
+        );
+    }
+
+    @Override
+    public void registerEvent(PaymentEvent event) {
+        this.events.add(event);
```

Comment:
```suggestion
       events.add(event);
```
pierdółka, ale tutaj `this` nie jest konieczne 😉